bnirc.1
.TH "bnirc" 1 June 11, 2005 "bnirc" "bnirc"

.SH SYNOPSIS
	bnirc [ -s startup file ] [ -p profile ] [ -h ]

.SH DESCRIPTION
	bnirc is an irc client

.SH OPTIONS
	-s startup file:
		specify a startup file, default is ~/.bnirc, this file just contains commands line-by-line to run on startup


	-p profile:
		default profile is 'default', also available is 'nocurses'

.SH NOTES
	none

.SH BUGS
	none known

.SH AUTHOR
	Brandon Niemczyk

.SH SEE ALSO



ChangeLog
2005-01-26  maccorin  <maccorin@maccorin.homelinux.org>

	* src/poll.c, include/irc.h, configure.ac:
	check errno after a poll before reporting a BUG() [because EINTR is ok].

2005-01-25  maccorin  <maccorin@maccorin.homelinux.org>

	* plugins/io_ncurses/io_ncurses.c, plugins/script/script.c, configure.ac:
	now takes CPPFLAGS from environment on ./configure and adds to them, instead
	of writing over them

	script now builds on my sparc box, although it is still failing to execute functions

	* src/plugin.c, include/irc.h, src/bnirc.c, src/io.c, src/parse_input_string.c, ChangeLog:
	cleaned up warnings on FBSD-sparc64

	* src/user.c: privmsg_print now calls ctcp_in_hook when approriate

2005-01-25  maccorin  <maccorin@maccorin.homelinux.org>

	* src/user.c: privmsg_print now calls ctcp_in_hook when approriate

2005-01-23  maccorin  <maccorin@maccorin.homelinux.org>

	* src/support.c: fixed fuckinggoogleit.com links

	* src/Makefile.am, src/irc.c, src/networkio.c, src/parse_input_string.c, src/plugin.c, src/support.c, include/plugin.h:
	various bug fixes
	added deps field to plugin_t
	moved networkio.c to more sensible name irc.c
	fixed bnirc_getline() to return error on EOF like GNU's does

2005-01-21  maccorin  <maccorin@maccorin.homelinux.org>

	* plugins/server_strings/server_strings.c:
	fixed bug where nicks stayed around in tab complete

	* plugins/io_ncurses/io_ncurses.c:
	fixed the bug where the lines cut off at the original size of the term instead of
	the real size

	* src/away.c, ChangeLog, common.tmplt: updated changelog
	added display_away option

2005-01-21  maccorin  <maccorin@maccorin.homelinux.org>

	* plugins/io_ncurses/io_ncurses.c:
	added some new color schemes to io_ncurses info line

	* src/bnirc.c, src/user.c, buildrelease.sh: fixed bug in recv_nick

2005-01-20  maccorin  <maccorin@maccorin.homelinux.org>

	* plugins/server_strings/server_strings.c:
	keeps user state after they quit if flags are set now

	* include/config.h.in: fixing build system...

	* scripts/Makefile.in, src/Makefile.in, plugins/rserver/Makefile.in, plugins/script/Makefile.in, plugins/server_strings/Makefile.in, include/Makefile.in, plugins/Makefile.in, plugins/ctcp/Makefile.in, plugins/dcc/Makefile.in, plugins/debug/Makefile.in, plugins/hello/Makefile.in, plugins/io_ncurses/Makefile.in, plugins/irc_input/Makefile.in, configure, configure.ac, Makefile.in:
	removed configure script and Makefile.in files from cvs

	* plugins/Makefile.am, plugins/Makefile.in, plugins/script/script.c, scripts/startup.sl, default.tmplt, include/list.h, nocurses.tmplt, common.tmplt:
	scripting is fixed [including festival stuff]

	* include/config.h.in, plugins/io_ncurses/io_ncurses.c, plugins/script/script.c, scripts/startup.sl, src/parse_server_string.c, common.tmplt, configure:
	now has the hide_info_line option
	as well as the /sinfoc command to set the color of your info line

2005-01-19  maccorin  <maccorin@maccorin.homelinux.org>

	* plugins/io_ncurses/io_ncurses.c, configure.ac:
	now uses default colors instead of black/white (although the api still assumes BLACK/WHITE)

	* plugins/io_ncurses/io_ncurses.c:
	fixed resizing so it doesn't kill all the text

	* src/user.c, common.tmplt: fixed /queue command

	* src/user.c:
	that's it... i'm done coding for today... see the last 2 commits... god i'm
	having issues

	* src/user.c:
	stupid mistakes.... fixes both handle_away_msg call and highlighting issues....

	* src/user.c:
	forgot to add handle_away_msg() call when i wrote the new privmsg code

	* src/bnirc.c, src/connect.c, src/options.c, src/support.c, include/irc.h, configure.ac:
	mostly cleanups, especially make_argv and make_max_argv
	the sigsegv handler now displays variables as well

	* src/poll.c: fixed stupid typo in comment

	* src/networkio.c, src/user.c, configure.ac, configure:
	fixed irc_nick going to wrong server on reconnect
	check for ncurses then curses in configure.ac added (why was it commented out?)

2005-01-18  maccorin  <maccorin@maccorin.homelinux.org>

	* src/user.c:
	the ACTION keyword on ctcp actions shouldn't be printed anymore

2005-01-17  maccorin  <maccorin@maccorin.homelinux.org>

	* src/bnirc.c, ChangeLog:
	fixed build error on non-linux systems where print_backtrace() was not defined

	* src/bnirc.c:
	added signal handler for SIGSEGV that prints out a backtrace

2005-01-16  maccorin  <maccorin@maccorin.homelinux.org>

	* ChangeLog: fixed changelog... it was all fscked up

	* ChangeLog: dummy-op

	* plugins/server_strings/server_strings.c, src/networkio.c, src/user.c, ChangeLog:
	fixes a stupid segfault from actions...

	* ChangeLog, src/user.c:
	using auto-update for changelogs now, added strlen() check in action_print() so nasty irc servers can't exploit us...

	* plugins/irc_input/irc_input.c, plugins/server_strings/server_strings.c, src/channel.c, src/io.c, src/networkio.c, src/support.c, src/user.c, include/irc.h, include/user.h, plugins/ctcp/ctcp.c:
	got rid of that ugly channel_msg() function, and replaced it with user->privmsg_print()

	* src/channel.c, src/networkio.c, include/channel.h:
	added default nick because some people were getting confused...
	now rejoins channels on a reconnect

	* src/support.c: sync

2005-01-15  maccorin  <maccorin@maccorin.homelinux.org>

	* src/channel.c: sync

	* src/networkio.c, src/support.c, include/irc.h:
	added bnirc_zmalloc [support.c] and added a malloc() macro [irc.h], so now
	all calls to malloc are guaranteed to return a zero-filled region of memory

	* src/networkio.c, src/readme.c:
	fixed reconnects (i think... haven't died yet, so it's not tested)

	* sample.bnirc: updated to use new auto-made backup nicks

	* README: updated to point to wiki

2005-01-14  maccorin  <maccorin@maccorin.homelinux.org>

	* src/connect.c, src/networkio.c, src/support.c, plugins/irc_input/irc_input.c:
	added auto-reconnect to the new connection interface

	* plugins/server_strings/server_strings.c: added timestamp to joins


src/bnirc.c
/*
 * Copyright (C) 2004 Brandon Niemczyk
 * 
 * DESCRIPTION:
 *
 * CHANGELOG:
 * 
 * LICENSE: GPL Version 2
 *
 * TODO:
 *
 */

#ifndef NDEBUG
#ifndef __cplusplus
extern int printf(const char *, ...);
#define DPRINT printf
#else
extern "C" int printf(const char *, ...);
#define DPRINT ::printf
#endif
#else
#define DPRINT(x,...)
#endif

#ifndef NULL
#define NULL ((void *)0)
#endif

#include <locale.h>
#include "irc.h"

static char *profile = STARTUPDIR "default";
static char *startup = ".bnirc";
static int default_startup = 1;

/**********************************************
 * valid options are:
 *  -p <profile name>
 *  -s <startup file>
 *  -h
 **********************************************/
static int do_options(int argc, char *argv[])
{
	int c;
    setlocale(LC_ALL, "");

	while (1) {
		c = getopt(argc, argv, "p:s:h");

		if (c == -1)
			break;

		switch (c) {
		case 'p':
			profile =
			    malloc(strlen(STARTUPDIR) + strlen(optarg) + 1);
			sprintf(profile, STARTUPDIR "%s", optarg);
			break;
		case 's':
			startup = malloc(strlen(optarg) + 1);
			strcpy(startup, optarg);
			default_startup = 0;
			break;
		case 'h':
			printf
			    ("usage: %s [-p <profile>] [-s <startup file>] [-h]\n",
			     argv[0]);
			exit(EXIT_SUCCESS);
		}
	}

	return 0;
}

#ifdef HAVE_BACKTRACE
#ifdef backtrace
#undef backtrace
#endif
/************************************************************************************************
 * i check for the backtrace_symbols function in my configure.ac with:
 *    AC_CHECK_FUNC(backtrace_symbols, [AC_DEFINE(HAVE_BACKTRACE, 1, [backtrace_symbols check])])
 ************************************************************************************************/
static void print_backtrace(int fd)
{
	const char start[] = "BACKTRACE ------------\n";
	const char end[] = "----------------------\n";

	void *bt[128];
	int bt_size;
	char **bt_syms;
	int i;

	bt_size = backtrace(bt, 128);
	bt_syms = backtrace_symbols(bt, bt_size);
	/* stderr is file descriptor 2 */
	write(fd, start, strlen(start));
	for (i = 1; i < bt_size; i++) {
		write(fd, bt_syms[i], strlen(bt_syms[i]));
		write(fd, "\n", 1);
	}
	write(fd, end, strlen(end));
}
#else
#define print_backtrace(x)
#endif

/**********************************************
 * SIGSEGV handler
 **********************************************/

static void _sigsegv_handler(int sig)
{
	unload_all_plugins();
	char msg[] =
	    "bnIRC has triggered a fatal error, if\n"
	    "you could make a bug report at:\n"
	    "http://sourceforge.net/tracker/?atid=698780&group_id=123832&func=browse\n"
	    "it would be very helpful\n\n"
	    "remove any security-sensitive information and\n"
	    "replace it with [removed for security]\n"
	    "or something similar to let me know\n" "Thank you.\n";

	write(2, msg, strlen(msg));
	if(current_input_string) {
		write(2, "processing: ", strlen("processing: "));
		write(2, current_input_string, strlen(current_input_string));
		write(2, "\n", 1);
	}

	if(current_server_string) {
		write(2, "processing: ", strlen("processing: "));
		write(2, current_server_string, strlen(current_server_string));
		write(2, "\n", 1);
	}
	print_backtrace(2);
	dump_options_fd(2);
	exit(EXIT_FAILURE);
}

static void stdin_poll_func(void *ignoreme)
{
	char *buf = io_get_input();
	if(buf) {
		plugin_input_hook(buf);
		parse_input_string(buf);
	}
}

int main(int argc, char *argv[])
{
	char *ts;
	irc_poll_t stdin_poll = {
		.callback = stdin_poll_func,
		.fd = fileno(stdin),
		.in = stdin
	};

	register_poll(&stdin_poll);
	// signal(SIGCHLD, SIG_IGN);
	signal(SIGSEGV, _sigsegv_handler);
	do_options(argc, argv);


	use_comments = 1;

	cio_out(PACKAGE "-" VERSION " - smack my BitchX up\n");
	cio_out("email bug reports to: bniemczyk@gmail.com\n");
	ts = timestamp();
	cio_out("starting at %s\n", ts);
	free(ts);

	/**********************************************
	 * begin the polling thread
	 **********************************************/
	// start_polling();

	char *buf = NULL;
	FILE *init_f = NULL;
	char *home = getenv("HOME");
	char *init_fname = NULL;
	size_t tmp;

	/**********************************************
	 * this is need to pre-initialize /set
	 **********************************************/
	if (!bool_option("verbose")) {
		cio_out("setting up /set command\n");
	}

	/**********************************************
	 * parse the profile and ~/.bnirc if it exists
	 **********************************************/
	if (home) {

		cio_out("HOME: %s\n", home);
		tmp = strlen(home);

		init_fname = malloc(strlen(profile) + 1);
		sprintf(init_fname, profile);
		init_f = bnirc_fopen(init_fname, "r");
		if (init_f) {
			tmp = 0;
			while (bnirc_getline(&buf, &tmp, init_f) != -1) {
				plugin_input_hook(buf);
				parse_input_string(buf);
			}
			fclose(init_f);
			cio_out("%s parsed\n", init_fname);
		} else {
			io_colored_out(USER_RED,
				       "%s cannot be opened or doesn't exist\n",
				       init_fname);
		}

		if (default_startup) {
			init_fname = malloc(strlen(home) + strlen(startup) + 1);
			sprintf(init_fname, "%s/%s", home, startup);
		} else {
			init_fname = startup;
		}
		init_f = bnirc_fopen(init_fname, "r");
		if (init_f) {
			tmp = 0;
			while (bnirc_getline(&buf, &tmp, init_f) != -1) {
				plugin_input_hook(buf);
				parse_input_string(buf);
			}
			fclose(init_f);
			cio_out("%s parsed\n", init_fname);
		} else {
			io_colored_out(USER_RED,
				       "%s cannot be opened or doesn't exist\n",
				       init_fname);
		}

	} else {
		io_colored_out(USER_RED, "no home directory found!\n");
	}

	/**********************************************
	 * main loop
	 **********************************************/
	use_comments = 0;
	while (!irc_shutdown) {
		/***************************************************************
		 * run loops calls the polling func, which sleeps,
		 * so we don't have to do that in here, we are nice
		 * anyways ;)
		 ***************************************************************/
		run_loop_hooks();
	}

	/**********************************************
	 * do some cleaning up
	 **********************************************/
	unload_all_plugins();

	return EXIT_SUCCESS;
}
